### ISSUE TITLE
Show that finite ordinals are natural numbers in iset.mm

### ISSUE THREAD

**Issue by jkingdon** [metamath/set.mm#2203]
There are a number of theorems, most of which which exist in set.mm, but which are missing from iset.mm, one or more of which is needed to get cardinality of finite sets ~~and http://us.metamath.org/mpeuni/df-hash.html unblocked (and thus allow resuming work on #2167 ):~~ _(Update Sep 2022: we defined https://us.metamath.org/ileuni/df-ihash.html another way and were able to prove most of the set size theorems from set.mm, so the topic of finite ordinals being natural numbers is still unresolved, but not blocking development of summation)._

* http://us.metamath.org/mpeuni/onomeneq.html
* http://us.metamath.org/mpeuni/onfin2.html
* http://us.metamath.org/mpeuni/ficardom.html
* http://us.metamath.org/mpeuni/cardnn.html
* `( ( A e. On /\ B e. _om /\ A ~~ B ) -> A e. _om )`

As far as I have been able to tell most/all of these are sufficient to imply the others, and in fact I've encountered similar difficulties in proving a variety of them.

The basic approaches I have tried are some variation of:

* Use http://us.metamath.org/ileuni/findcard2sd.html or one of the other findcard* theorems on the finite ordinal. Usually the difficulty here is that you can add an element to a finite ordinal (or subtract one in the case of `findcard`), but I haven't found a way to identify which element would result in that also giving us an ordinal. That sort of thing is solved for natural numbers, because we have the successor (or the predecessor) but there is nothing order-preserving about the 1-1 and onto mapping we get from `n ~~ A`. It would help if we could say a finite ordinal is either `(/)` or the successor of some ordinal, but I haven't been able to prove that.
* Use http://us.metamath.org/ileuni/finds.html on the natural number and do something along the lines of showing, one induction step at a time, that the elements of the natural number are found in the finite ordinal. This has a roughly similar problem: we can identify the next natural number to look at, but it isn't clear how to map that over to the finite ordinal and show that the finite ordinal contains each natural number, or can produce an ordinal at each step, or the like.
* Is there some kind of special case of ordinal trichotomy (just for finite ordinals/sets) or anything similar which can get us out of this? I've found a few cool things, for example http://us.metamath.org/ileuni/fisbth.html , http://us.metamath.org/ileuni/fidceq.html (which I guess can be stated as "finite sets are discrete"), http://us.metamath.org/ileuni/php5fin.html (and a few other pigeonhole principle variations), and http://us.metamath.org/ileuni/diffifi.html (which is a modified version of http://us.metamath.org/mpeuni/diffi.html which we can't have as-is). However, it isn't clear to me whether these, or other similar results we might be able to prove, add up to being able to prove `onomeneq` and the like.
* Of course when a lot of attempts seem to hit obstacles I consider whether it is telling me that it can't be done intuitionistically. Which I won't know for sure unless we prove it or prove it implies excluded middle (or a similar impossibility proof), but on the face of it `A e. Fin` would seem to rule out ` A ` being a "set of undecided size" like ` {𝑧 ∈ {∅} ∣ 𝜑} ` from http://us.metamath.org/ileuni/ordtriexmid.html . There's nothing especially magic about keeping http://us.metamath.org/ileuni/df-card.html (since I have very low expectations for cardinals beyond finite sets), but if this definition (taken from set.mm) can be made to work, I guess that it is better than changing it just because we haven't figured out what does and does not work?

**Comment by digama0** [comment]
I think your first proof attempt is on the right track (and I believe this theorem to be true). What you need, to get around the fact that taking an element out of the middle of an ordinal makes it not an ordinal anymore, is the ordinal collapse function, which goes by the name `OrdIso` in set.mm, together with its main theorem http://us.metamath.org/mpeuni/ordtype.html .

Lemma: If `A` is a finite set of ordinal numbers, `A ~~ n`, then its order type is `n` (a natural number). If `A ~~ (/)` then it is empty and the order type is `0`; and if `A ~~ suc n` then let `f : suc n --> A` be a bijection, and let `B = (f " n)`. Then `B` is a subset of `A`, and a set of ordinal numbers, and `B ~~ n`, so by the inductive hypothesis `n` is the order type of `B`. Now we have a lemma that `(B u. {x})` where `x e/ B` (still dealing with subsets of `On`) has order type `suc n`, which goes by determining where `x` is inserted in the sequence `B`. (I can give more details later on this point.) And then we are done, setting `x` to ``(f ` n)``, since ``(f " n) u. (f ` n) = A``.

Now we have the theorems as follows:

1. If `A e. On` and `n e. om` and `A ~~ n`, then the order type of `A` is `A` so the lemma applies and `A = n` and `A e. om`.
2. `om = ( On i^i Fin )`: the forward direction is obvious, and a finite ordinal satisfies the conditions for (1) so it is in `om`.
3. If `n e. om` then ``(card ` n)`` is a finite ordinal so it is a natural number by (2), and ``(card ` n) ~~ n`` so ``(card ` n) = n`` by (1).
4. If `A` is finite, `A ~~ n`, then ``(card ` A) = (card ` n) = n`` by (3), so ``(card ` A) e. om``.

**Comment by jkingdon** [comment]
> What you need . . . is the ordinal collapse function

Ah. Buckle your seat belts, we are in for quite a ride making this a reality. As far as I could tell/guess based on what I've looked over so far, this can work, but there are a bunch of parts to it:

* Might as well add `Fr` to start the process of reducing ax-setind usage as described at https://github.com/metamath/set.mm/pull/1044#issuecomment-921354430 . This is mostly because I know it will make @digama0 happy, but also to reduce unnecessary divergence from set.mm. The connection to ordinal collapse is that many of the definitions and theorems related to `OrdIso` in set.mm reference `Fr`, so it seems better to at least include that in the relevant places, whether or not we go very far down the path of reducing `ax-setind` usage just yet. I'll make a follow up issue or pull request for `Fr` rather than elaborating further here.
* I assume we want to define `We` to mean a trichotomous order, a la http://us.metamath.org/mpeuni/dfwe2.html or something similar. At least, natural numbers are trichotomous, and so are finite sets if we order them according to the 1-1,onto map to a natural number.
* Keeping the `if` statement in http://us.metamath.org/mpeuni/df-oi.html seems harmless even if the condition isn't decidable, although I guess it is vestigial without excluded middle. The condition will be true in the http://us.metamath.org/mpeuni/ordtype.html case. Presumably we can also add a `(𝑅 We 𝐴 ∧ 𝑅 Se 𝐴)` condition to theorems like http://us.metamath.org/mpeuni/oicl.html without removing anything we are going to need.
* I didn't get as far as trying to figure out whether the function ` G ` from http://us.metamath.org/mpeuni/dfoi.html satisfies the additional condition of, for example, http://us.metamath.org/ileuni/tfri1d.html .
* In general, I haven't yet read through the whole http://us.metamath.org/mpeuni/ordtype.html proof including the various lemmas. So if there are additional dragons other than the ones mentioned above, I haven't found them yet.
* Having said all that, it still seems like a good idea to stick as close to set.mm as we can even if some of the results won't be as powerful as what is possible with excluded middle.

**Comment by digama0** [comment]
> I assume we want to define We to mean a trichotomous order, a la http://us.metamath.org/mpeuni/dfwe2.html or something similar. At least, natural numbers are trichotomous, and so are finite sets if we order them according to the 1-1,onto map to a natural number.

I'm not sure about this. My guess is that we want `We` to mean something equivalent to the property that ordinals have with `e.`, and ordinals are not trichotomous. The current definition says that a set is an ordinal if it is transitive and all its elements are transitive. For a general relation, I think we just need that it is set-like, transitive, and satisfies the equivalent of ax-setind:

* For any `x e. A`, the set `down(x) = { y e. A | y R x }` is a set
* If `x,y,z e. A` and `x R y R z` then `x R z`
* If for all `x e. A`, `{ y e. A | y R x } C_ s` implies `x e. s`, then `A C_ s`

The last bullet is probably what we want to call `R Fr A`. It suffers from the same issue as ax-setind regarding proper classes though; I don't know if there are any easy fixes, but it might help to have a version `FrFor R A S` with `s` being a parameter, so that it can be a proper class (and then proofs can show that this holds for all classes `S`, and applications will instantiate particular class parameters for `S`). When `A` is a set, of course, this is not needed.

**Comment by jkingdon** [comment]
> My guess is that we want `We` to mean . . . set-like, transitive, and satisfies the equivalent of ax-setind:

Hmm, OK, if those things suffice, sure. The first bullet is already in iset.mm as http://us.metamath.org/ileuni/df-se.html .

> `R Fr A` . . . proper classes though; I don't know if there are any easy fixes, but it might help to have a version `FrFor R A S` with `s` being a parameter, so that it can be a proper class (and then proofs can show that this holds for all classes `S`, and applications will instantiate particular class parameters for `S`). When `A` is a set, of course, this is not needed.

I was going to make another issue for `Fr` but we are already talking about it so I'll respond here. I've mostly been working with

```
    df-frind $a |- ( R Fr A <->
        A. s ( A. x e. A ( A. y ( y R x -> y e. s ) -> x e. s )
          -> A C_ s ) ) $.
```

and working on proving http://us.metamath.org/mpeuni/frirr.html . Seems to be working - using the proof from https://github.com/metamath/set.mm/issues/647#issuecomment-441550666 which is basically also the proof at http://us.metamath.org/ileuni/elirr.html . The issues so far are needing to add `A e. V` to what I am proving (as you note, the above definition is for when `A` is a set) and changing `A. x` to `A. x e. A` as shown above.

I don't know whether to finish the proof with that definition or move on to dealing with `A` possibly being a proper class.

What is the difference between `A` and `S` in the proposed `FrFor`? And how does that differ from taking http://us.metamath.org/ileuni/setindel.html and changing the first `e.` to `R`? I'll admit I still sometimes get lost on what our choices are but I try to focus by asking myself "how does this let us prove frirr or whatever its equivalent ends up being?"

**Comment by digama0** [comment]
> What is the difference between A and S in the proposed FrFor? And how does that differ from taking http://us.metamath.org/ileuni/setindel.html and changing the first e. to R?

It is exactly this. More precisely, the proposed definition is (slightly amending `A. y` to `A. y e. A` since the order classes don't assume `x R y -> x e. A` or `x R y -> y e. A`):

    df-frfor $a |- ( FrFor R A S <->
        ( A. x e. A ( A. y e. A ( y R x -> y e. S ) -> x e. S )
          -> A C_ S ) ) $.

In other words, it is just the body of the forall, so that `R Fr A <-> A. s FrFor R A s`. This allows properties to be proven that require `A` and `S` to both be proper classes, avoiding the `A e. V` assumption that you mention.

For many applications (including ordtype, I think), you don't need `A` to be a proper class, but there are enough theorems on the border that I would suggest keeping the proper class version in mind.

**Comment by jkingdon** [comment]
> the proposed definition is . . . 

Ah, cool. This is now in pull request form at #2209 and seemed to go fairly smoothly. Once we changed those two `A. ?` to `A. ? e. A ` and decided to not spend a huge amount of focus on ` A ` being a proper class, the things which had been tripping me up went away.

> For many applications (including ordtype, I think), you don't need `A` to be a proper class, but there are enough theorems on the border that I would suggest keeping the proper class version in mind.

Ah, OK. My head was full of http://us.metamath.org/mpeuni/onfr.html (where ` A ` is a proper class) and http://us.metamath.org/mpeuni/zfregfr.html (where ` A ` might be a proper class). But you make a good point, that our current goal is ordtype.

**Comment by jkingdon** [comment]
Here's my latest attempt at onomeneq (in case us2 has not been updated when you read this, some of the theorems referenced are from https://github.com/metamath/set.mm/pull/2222 ):

First note that ` F Isom _E , _E ( A , B ) ` is sufficient to show ` A = B ` by http://us2.metamath.org/ileuni/ordiso2.html

We will prove ` ( _E We A -> E. n e. _om E. f f Isom _E , _E ( A , n ) ) ` by http://us.metamath.org/ileuni/findcard.html induction on A. The base case is easy. The induction step amounts to taking an order-isometry between ` ( y \ { v } ) ` and ` k ` and turning it into an order-isometry between ` y ` and ` suc k ` . We need to add a mapping between ` v ` and ` k `. Although this is all peachy on the ` k ` side, on the ` y ` side we have the problem that ` v ` might be in the middle of ` y ` rather than at the end. That ` ( y \ { v } ) ` is well-ordered follows from http://us2.metamath.org/ileuni/wessep.html but we still need some flavor of ordinal collapse (at least for this particular case) to make this work.

In other words, I'm not sure this advances matters beyond our previous discussion here and at #2212

**Comment by digama0** [comment]
I still think the ordtype proof can be salvaged here. The basic idea is that the general proof only holds when we know that all the iota applications in the definition of the `OrdIso` function exist, and for a finite set of size `n` we should be able to show that exactly `n` of the searches succeed (and the `n+1`-th search for the next largest element fails).

The simplest case of this (which is also the general case) is when `n = 1`. Suppose we know that `A` is inhabited. Does it have an `_E`-minimal element, which we can uniquely define using an `iota` expression? The expression from OrdIso is something like this: `( iota_ x e. A A. y e. A x C_ y )`. The predicate here is clearly unique, and I think existence follows from `Tr A`.

Repeating this `n` times, we can prove that `f(0)` (where `f = OrdIso(A, n)`) is the smallest element in `A`, `f(1)` is the smallest element in `{x e. A| f(0) < x}` (which has size `n-1`), so we are building up a strict inequality chain. Repeating this argument `n` times results in an increasing list of elements of `A` of length `n`, and the induction hypothesis should give that the set of untouched elements has size 0 and hence is empty, so this is an order isomorphism.

**Comment by jkingdon** [comment]
I think I'm several steps behind you on most of this, so let me follow up on just one part:

> something like this: `( iota_ x e. A A. y e. A x C_ y )`. The predicate here is clearly unique, and I think existence follows from `Tr A`

So something like this?

```
    trminss $p |- ( ( Tr A /\ E. w w e. A ) -> E. x e. A A. y e. A x C_ y ) $=
      ? $.
```

I suspect this may be provable (at least, it doesn't seem like the set ` A ` at http://us.metamath.org/ileuni/regexmidlem1.html could be shown to be transitive), although I'm not sure I see how. I looked at http://us.metamath.org/ileuni/trint0m.html which didn't seem like quite what I was looking for.

**Comment by digama0** [comment]
Yes, the theorem statement is correct, and trint0m looks like it's on the right track - `|^| A` looks like it should be the necessary witness. It must exist when A is inhabited, and it is also clearly a subset of every element of A, so the only remaining part is to prove that `|^| A e. A`, which relies on this being a set of ordinals.

I'm pretty sure you have to use well foundedness / ax-setind here somewhere, because the theorem is essentially a form of the axiom of foundation. Is there an exmid-style counterexample for `|^| A e. A`?

**Comment by jkingdon** [comment]
> Is there an exmid-style counterexample for `|^| A e. A`?

Yes, http://us.metamath.org/ileuni/onintexmid.html so being a set of ordinals isn't enough by itself.

As a style point, do you think I should change the hypothesis of onintexmid to ∀w ((w ⊆ On ∧ ∃𝑥𝑥 ∈ w) → ∩ w ∈ w) ? When I wrote it the way I did I was worried about set versus class but the counterexample is a set so either way of writing it would work and the way it is written now I find a bit jarring. Like every time I look at it I need to remember that the thing it is a counterexample for is in the comment rather than directly in the hypothesis.

**Comment by digama0** [comment]
Ooh, that counterexample is bad, really bad. In the focused special case, we have a set with two elements that we can't order. This is a direct contradiction to the theorem we want to prove, since if we have an order isomorphism from `{ y , z }` to `{ 0 , 1 }` then `y e. z`, and vice versa for the other bijection. So there is no ordinal collapse function for this set.

Re: style point, I agree that it would be better to use `∀w ((w ⊆ On ∧ ∃𝑥𝑥 ∈ w) → ∩ w ∈ w)`, or even `((w ⊆ On ∧ ∃𝑥𝑥 ∈ w) → ∩ w ∈ w)` if it's not too confusing.

**Comment by jkingdon** [comment]
> if we have an order isomorphism from `{ y , z }` to `{ 0 , 1 }` then `y e. z`, and vice versa for the other bijection. So there is no ordinal collapse function for this set.

Well, the fact we haven't used yet is that (at least in a case like onomeneq or onfin2) `{ y , z }` is an ordinal. For equality we have http://us2.metamath.org/ileuni/fidceq.html but is there anything at all along those lines for `e.`?

I mean, if it can't be done that would be a fine answer too, but I haven't been able to find a counterexample either, at least so far. For example suppose y is `{x ∈ {∅, {∅}} ∣ 𝜑}` (taken from the http://us2.metamath.org/ileuni/ordtri2or2exmid.html proof) and z is ` {∅} `. Well the hypothesis that `{ y , z }` is transitive would force y into the `∅` case rather than the `{∅, {∅}}` case.

> Re: style point, I agree that it would be better to use `∀w ((w ⊆ On ∧ ∃𝑥𝑥 ∈ w) → ∩ w ∈ w)`, or even `((w ⊆ On ∧ ∃𝑥𝑥 ∈ w) → ∩ w ∈ w)` if it's not too confusing.

Ah, we are thinking the same way. What I now have on the https://github.com/jkingdon/set.mm/tree/iset-min branch is:

```
    onintexmid.onint $e |- ( ( y C_ On /\ E. x x e. y ) -> |^| y e. y ) $.
    $( If the intersection (infimum) of an inhabited class of ordinal numbers
       belongs to the class, excluded middle follows.  The hypothesis would be
       provable given excluded middle.  (Contributed by Mario Carneiro and Jim
       Kingdon, 29-Aug-2021.) $)
    onintexmid $p |- ( ph \/ -. ph ) $=
      . . .
```

**Comment by digama0** [comment]
The theorem statement doesn't really make sense if you just add `Tr A`, because the empty set is always a member of a nonempty transitive set of ordinals, so the intersection is empty. Note that the sets we want to apply this construction to are definitely *not* transitive, at least classically; assuming it has a minimum element, any member of that minimum will not be in the set.

**Comment by jkingdon** [comment]
> The theorem statement doesn't really make sense if you just add `Tr A`, because the empty set is always a member of a nonempty transitive set of ordinals, so the intersection is empty. Note that the sets we want to apply this construction to are definitely _not_ transitive, at least classically; assuming it has a minimum element, any member of that minimum will not be in the set.

Hmm, we seem to both have been a bit confused about this earlier on this thread. Perhaps we meant ` _E We A ` instead of `Tr A`? If so I can answer that the branch now contains a proof for:

```
  ${
    $d ph u w z $.  $d ph u x y z $.
    reg3exmid.1 $e |- ( ( _E We z /\ E. w w e. z ) ->
        E. x e. z A. y e. z x C_ y ) $.
    $( If any inhabited set satisfying ~ df-wetr for ` _E ` has a minimal
       element, excluded middle follows.  (Contributed by Jim Kingdon,
       3-Oct-2021.) $)
    reg3exmid $p |- ( ph \/ -. ph ) $=
      . . .
```

**Comment by jkingdon** [comment]
Although the comment thread here went in a lot of directions, the current status:

* The problem of whether https://us.metamath.org/mpeuni/onomeneq.html can proved is still unsolved, with neither a proof nor a proof of impossibility
* Various theorems are proved, shown to imply excluded middle, and/or noted in mmil.thml as appropriate, 
* We have found other ways to prove most of the theorems we need related to finite sets, the set size function, and the summation notation. For this reason, I suspect we should close this issue and let us open more focused issues if we want to revisit.

**Comment by jkingdon** [comment]
To be explicit about why this is being closed, #3861 is a proof that if any finite ordinal is a natural number, excluded middle follows.

Development of iset.mm has proceeded along other lines, so many of the set.mm theorems about finite sets have been proved other ways (or are not possible). But we at least know that trying to use `onomeneq` as set.mm does is a dead end for iset.mm.
